nifi-nar-bundles/nifi-standard-services/nifi-lookup-services-bundle/nifi-lookup-services/src/main/java/org/apache/nifi/lookup/DistributedMapCacheRecordLookupService.java

```diff
@@ -0,0 +1,183 @@
+ * limitations under the License.
+ */
+
+package org.apache.nifi.lookup;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.ByteArrayInputStream;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.HashMap;
+import java.io.InputStream;
+
+import org.apache.nifi.annotation.documentation.CapabilityDescription;
+import org.apache.nifi.annotation.documentation.Tags;
+import org.apache.nifi.annotation.lifecycle.OnEnabled;
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.components.Validator;
+import org.apache.nifi.controller.AbstractControllerService;
+import org.apache.nifi.controller.ConfigurationContext;
+import org.apache.nifi.distributed.cache.client.Deserializer;
+import org.apache.nifi.distributed.cache.client.DistributedMapCacheClient;
+import org.apache.nifi.distributed.cache.client.Serializer;
+import org.apache.nifi.distributed.cache.client.exception.DeserializationException;
+import org.apache.nifi.distributed.cache.client.exception.SerializationException;
+import org.apache.nifi.expression.ExpressionLanguageScope;
+import org.apache.nifi.serialization.record.Record;
+import org.apache.nifi.serialization.RecordReader;
+import org.apache.nifi.serialization.RecordReaderFactory;
+import org.apache.nifi.util.Tuple;
+import org.apache.nifi.lookup.RecordLookupService;
+
+@Tags({ "lookup", "enrich", "key", "value", "map", "cache", "distributed" })
+@CapabilityDescription("Allows to choose a distributed map cache client to retrieve the value associated to a key as Record. "
+        + "The coordinates that are passed to the lookup must contain the key 'key'.")
+public class DistributedMapCacheRecordLookupService extends AbstractControllerService implements RecordLookupService {
+
+    private static final List<Charset> STANDARD_CHARSETS = Arrays.asList(StandardCharsets.UTF_8, StandardCharsets.US_ASCII, StandardCharsets.ISO_8859_1, StandardCharsets.UTF_16,
+            StandardCharsets.UTF_16LE, StandardCharsets.UTF_16BE);
+
+    private static final String KEY = "key";
+    private static final Set<String> REQUIRED_KEYS = Stream.of(KEY).collect(Collectors.toSet());
+    private volatile RecordReaderFactory readerFactory;
+
+    private volatile DistributedMapCacheClient cache;
+    private volatile static Charset charset;
+    private final Serializer<String> keySerializer = new StringSerializer();
+    private final Deserializer<byte[]> valueDeserializer = new CacheValueDeserializer();
+
+    public static final PropertyDescriptor PROP_DISTRIBUTED_CACHE_SERVICE = new PropertyDescriptor.Builder().name("distributed-map-cache-service").displayName("Distributed Cache Service")
```

Comment:
One method call per line would make these easier to read. That's a general formatting structure we tend to use.